### PR TITLE
feat: write buffer ingestion metrics

### PR DESCRIPTION
**Pull request by crepererum**

### CODE REVIEW

File: server/src/db.rs [hunk 1]
```diff
@@ -144,6 +144,94 @@ pub enum Error {
...
+    domain: Arc<metrics::Domain>,
+}
+
+impl IngestMetrics {
+    fn new(domain: Arc<metrics::Domain>) -> Self {
+        Self { domain }
+    }
+
+    fn new_sequencer_metrics(&self, sequencer_id: u32) -> SequencerMetrics {
+        let labels = vec![KeyValue::new("sequencer_id", sequencer_id.to_string())];
```
Comment (mkmik) [inline server/src/db.rs:160]: Can we also have the database name as a label?

Comment (mkmik) [reply server/src/db.rs:160]: Ah. i see in the tests there is a db_name label. I found it hard to see which labels we actually set; that's the problem of our current metrics API

Comment (crepererum) [reply server/src/db.rs:160]: Should I add this to #2076?

Comment (mkmik) [reply server/src/db.rs:160]: Good idea, yes please

File: server/src/db.rs [hunk 2]
```diff
@@ -144,6 +144,91 @@ pub enum Error {
...
+        let last_min_ts = self.domain.register_gauge_metric_with_labels(
+            "last_min_ts",
+            None,
+            "Minimum unix timestamp of last write as unix timestamp in nanoseconds",
+            &labels,
+        );
+        let last_max_ts = self.domain.register_gauge_metric_with_labels(
+            "last_max_ts",
+            None,
+            "Maximum unix timestamp of last write as unix timestamp in nanoseconds",
```
Comment (mkmik) [inline server/src/db.rs:192]:
```suggestion
            "Maximum timestamp of last write as unix timestamp in nanoseconds",
```

File: server/src/db.rs [hunk 3]
```diff
@@ -144,6 +144,91 @@ pub enum Error {
...
+        let sequence_number_lag = self.domain.register_gauge_metric_with_labels(
+            "sequence_number_lag",
+            None,
+            "The difference between the last consumed sequence number (e.g. Kafka offset) and the last sequence number available",
+            &labels,
+        );
+        let last_min_ts = self.domain.register_gauge_metric_with_labels(
+            "last_min_ts",
+            None,
+            "Minimum unix timestamp of last write as unix timestamp in nanoseconds",
```
Comment (mkmik) [inline server/src/db.rs:186]:
```suggestion
            "Minimum timestamp of last write as unix timestamp in nanoseconds",
```

File: server/src/db.rs [hunk 4]
```diff
@@ -144,6 +144,91 @@ pub enum Error {
...
+        let last_sequence_number = self.domain.register_gauge_metric_with_labels(
+            "last_sequence_number",
+            None,
+            "Last consumed sequence number (e.g. Kafka offset)",
+            &labels,
+        );
+        let sequence_number_lag = self.domain.register_gauge_metric_with_labels(
+            "sequence_number_lag",
+            None,
+            "The difference between the last consumed sequence number (e.g. Kafka offset) and the last sequence number available",
```
Comment (mkmik) [inline server/src/db.rs:180]: should it be the difference of the last available - the last consumed?

**Review by mkmik — APPROVED** [review]
LG, minor a few minor comments.

One thing I'd like to expose (perhaps in another PR) is the last sequence number as seen by the writer process.
IIUC currently, the consumer is reporting the sequence number of the currently ingested entry and the lag w.r.t what *it* knows about where the last written sequence number is (that's the watermark, right?). The downside of this is that if for some reason (bug,...) the consumer is stuck at thinking that the watermark is not growing, the reported lag would lead us to think there is no lag.

In my past experience with pipeline systems I've usually exported the sequence numbers independently from the producers and consumers and delegated the lag computation to the monitoring system.

File: server/src/db.rs [hunk 5]
```diff
@@ -144,6 +144,91 @@ pub enum Error {
...
+        let bytes_read = self.domain.register_counter_metric_with_labels(
+            "read",
+            Some("bytes"),
+            "Bytes read from sequencer",
+            labels.clone(),
+        );
+        let last_sequence_number = self.domain.register_gauge_metric_with_labels(
+            "last_sequence_number",
+            None,
+            "Last consumed sequence number (e.g. Kafka offset)",
```
Comment (e-dard) [inline server/src/db.rs:174]: 👍

**Review by e-dard — APPROVED** [review]
This LGTM. I think as a follow-on PR we should expose a metric to track  the wall-clock lag when consuming messages. That is - how long did it take a produced message to be ingested by the query server.

File: server/src/db.rs (hunk 4 above)
Comment (crepererum) [reply server/src/db.rs]: fixed (clarified)

File: server/src/db.rs (hunk 3 above)
Comment (crepererum) [reply server/src/db.rs]: fixed

File: server/src/db.rs (hunk 2 above)
Comment (crepererum) [reply server/src/db.rs]: fixed

File: server/src/db.rs [hunk 6]
```diff
@@ -144,6 +144,94 @@ pub enum Error {
...
+        let bytes_read = self.domain.register_counter_metric_with_labels(
+            "read",
+            Some("bytes"),
+            "Bytes read from sequencer",
+            labels.clone(),
+        );
+        let watermark_iox = self.domain.register_gauge_metric_with_labels(
+            "watermark_iox",
+            None,
+            "High watermark of IOx (aka next sequence number that will be ingested)",
```
[the remark this answers is not in the document]
Comment (crepererum) [reply server/src/db.rs]: done

File: server/src/db.rs [hunk 7]
```diff
@@ -705,32 +797,116 @@ impl Db {
...
+                metrics
+                    .last_max_ts
+                    .set(max_ts.timestamp_nanos() as usize, &[]);
+            }
+
+            // maybe update sequencer watermark
+            // We are not updating this watermark every round because asking the sequencer for that watermark can be
+            // quite expensive.
+            if last_watermark_update
+                .map(|ts| ts.elapsed() > Duration::from_secs(60))
```
[the remark this answers is not in the document]
Comment (crepererum) [reply server/src/db.rs]: done

**Comment by crepererum** [comment]
> One thing I'd like to expose (perhaps in another PR) is the last sequence number as seen by the writer process.

Sure, we can add a whole set of metrics on the write side as well.

> This LGTM. I think as a follow-on PR we should expose a metric to track the wall-clock lag when consuming messages. That is - how long did it take a produced message to be ingested by the query server.

We can certainly do that (rdkafka provides a msg. timestamp). However that requires a bit more wiring since the sequenced entries currently don't carry a timestamp.

File: server/src/db.rs [hunk 8]
```diff
@@ -1279,6 +1455,71 @@ mod tests {
             tokio::time::sleep(Duration::from_millis(100)).await;
         }
 
+        // check: metrics
+        // We need to do that BEFORE shutting down the background loop because gauges would be dropped and resetted otherwise
+        let metrics = test_db.metric_registry;
+        metrics
+            .has_metric_family("ingest_write_buffer_requests_total")
```
[the remark this answers is not in the document]
Comment (crepererum) [reply server/src/db.rs]: done

File: server/src/db.rs (hunk 1 above)
Comment (crepererum) [reply server/src/db.rs:160]: done

**Pull request merged**